envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -31,40 +32,75 @@ class SnapshotUpdater(
         properties = properties
     )
 
-    fun start(changes: Flux<List<LocalityAwareServicesState>>): Flux<List<LocalityAwareServicesState>> {
-        // see GroupChangeWatcher
-        return Flux.combineLatest(
-            onGroupAdded,
-            changes,
-            BiFunction<Any, List<LocalityAwareServicesState>, List<LocalityAwareServicesState>> { _, states -> states }
-        )
-            .sample(properties.stateSampleDuration)
-            .publishOn(scheduler)
-            .doOnNext { states ->
-                versions.retainGroups(cache.groups())
-                updateSnapshots(states)
+    fun start(changes: Flux<List<LocalityAwareServicesState>>): Flux<UpdateResult> {
+        return Flux.merge(
+                services(changes),
+                groups()
+        ).doOnNext { result ->
+            val groups = if (result.action == Action.ALL_GROUPS) {
+                cache.groups()
+            } else {
+                result.groups
             }
-            .onErrorResume { e ->
-                meterRegistry.counter("snapshot-updater.services.updates.errors").increment()
-                logger.error("Unable to process service changes", e)
-                Mono.justOrEmpty(listOf())
+
+            versions.retainGroups(cache.groups())
+            groups.forEach { group ->
+                if (group.ads) {
+                    updateSnapshotForGroup(group, lastAdsSnapshot)
```

Comment:
It seems you've tried to skip the whole functional idiom here. I don't have a clear idea how to make this work yet, but that's very error prone, because you'll be changing schedulers along the way and possibly threads. The references to `lastXSnapshot` are not guarded in any way. Still, I wouldn't go for such an imperative approach here - if fluxes are here to stay - let's figure out a fluxy way of caching the snapshots.

Reply:
`because you'll be changing schedulers along the way and possibly threads` is not true but I get your point. See: https://github.com/allegro/envoy-control/pull/49#discussion_r357047001

Reply:
Also: Flux is a construct from project reactor. Project reactor is about building non blocking applications "based on the Reactive Streams Specification". Reactive streams specification does not mention anything about the programming style used: https://github.com/reactive-streams/reactive-streams-jvm#goals-design-and-scope. 

FP is at odds with mutating data but if we want to do caching then I think we have no choice. You can either do caching and mutate things or compute the value every time and impact the performance.

Reply:
Also2: If you really want to we can use memoization, but if you want to see what's underneath see: https://github.com/arrow-kt/arrow/blob/ce539f2c807683eb937190dfcdb76ddd2db75cad/modules/meta/arrow-meta/src/main/java/arrow/meta/internal/memoization.kt#L25 ;)

Reply:
Yes, it seems memoization would be the way to go, but I'd suspect the `Flux.cache` function achieves that. I'd check it or possible use a combination of a Flux of snapshots: `publishedLatestSnapshots` and a merge with: `.repeat().takeUntilOther(publishedLatestSnapshots)`

Reply:
I do not think Flux.cache works that way, but I might be wrong. It is used to cache last results to REPLAY them for a new subscriber, it does not provide a way to see previous values.

Consider the following code:
```
fun main() {
    val startFlux = Flux.interval(Duration.ofSeconds(1)).cache(2)

    startFlux.subscribe {
        println("${LocalDateTime.now()} $it")
    }

    Thread {
        Thread.sleep(5000)
        startFlux.subscribe {
            println("${LocalDateTime.now()} late comer $it")
        }
    }.start()

    Thread.sleep(10000)
}
```

The output is: 
```
2019-12-18T11:43:14.537 0
2019-12-18T11:43:15.522 1
2019-12-18T11:43:16.525 2
2019-12-18T11:43:17.524 3
2019-12-18T11:43:18.524 4
2019-12-18T11:43:18.524 late comer 3
2019-12-18T11:43:18.525 late comer 4
2019-12-18T11:43:19.522 5
2019-12-18T11:43:19.522 late comer 5
2019-12-18T11:43:20.523 6
2019-12-18T11:43:20.523 late comer 6
2019-12-18T11:43:21.526 7
2019-12-18T11:43:21.526 late comer 7
2019-12-18T11:43:22.527 8
2019-12-18T11:43:22.527 late comer 8
2019-12-18T11:43:23.522 9
2019-12-18T11:43:23.522 late comer 9
```

As you can see when a late comer comes in at 5s it gets cached values from a flux (3 and 4 - as we set history depth of 2).